### ISSUE TITLE
Replace deprecated mongodb $pushAll operator with $push + $each

### ISSUE THREAD

**Issue by pekebuda** [Automattic/mongoose#4455]
According to MongoDB [documentation](https://docs.mongodb.com/manual/reference/operator/update/pushAll/) `$pushAll` has been deprecated since version 2.4. However, it remains the operator Array#push translates into.  

On the assumption that is not a breaking change, I think it would be a convenient fix, since some implementation may not rely on `$pushAll`. In fact, for what it is worth, [Azure's own MongoDB implementation](https://azure.microsoft.com/en-gb/documentation/articles/documentdb-protocol-mongodb/) does not support `$pushAll` operator AFAIK .

**Comment by vkarpov15** [comment]
Getting off of pushAll is definitely a priority, but fair warning, mongoose does *not* support documentdb at the moment, so use it at your own risk

**Comment by lamuertepeluda** [comment]
This may mitigate the impact of Array.push relying on $pushAll. I tried it myself and seems to work.  

```javascript
myArray.push(myObject); //breaks on DocumentDB with Mongo API because of $pushAll
```

instead use

```javascript
myArray = myArray.concat([myObject]); //this uses $set so no problems
```


@vkarpov15 Supporting [DocumentDB with Mongo Protocol](https://azure.microsoft.com/en-us/documentation/articles/documentdb-protocol-mongodb/) (so not DocumentDB as it is) would be fantastic, even though I can understand that project priorities are on Mongo. Sadly Microsoft seem to have implemented just a fair subset of the latest Mongo API (3.x), so deprecated stuff that comes from 2.x may not work.

**Comment by vkarpov15** [comment]
In 4.6.4 you'll be able to use the `usePushEach` option in schemas: `new Schema({ arr: [String] }, { usePushEach: true });`

**Comment by knutole** [comment]
Started running into this problem on Mongo 3.6 and Mongoose 4.13.6, ` message: 'Unknown modifier: $pushAll',`

Suggested fix does not work:
```
var layerSchema = mongoose.Schema({
   (...) 
}, 
{ 
    usePushEach: true 
});

```

It's weird that this is still an issue with the updated libs?

**Comment by frguthmann** [comment]
Same as Knutole here, I'm creating models with mongoose and I don't know if / where I can plugin that fix.

**Comment by jonatanrinckus** [comment]
This suggestion solve the problem in mongodb 3.6 for me.

**Comment by silveur** [comment]
Same issue here using Mongo 3.6
Any push and save on array field generate the error: Unknown modifier: $pushAll
Using the latest Mongoose too...

**Comment by pakastin** [comment]
Workaround `usePushEach: true` seems to work for us. MongoDB 3.6 + `db.adminCommand({ setFeatureCompatibilityVersion: "3.6" })`

**Comment by grantgeorge** [comment]
@knutole what issue are you running into? This solution has worked for me. @frguthmann safest bet would be to add that line to all models if you're using MongoDB 3.6+.

**Comment by grantgeorge** [comment]
Mongo 3.6 was recently released and [removed support for $pushAll](https://docs.mongodb.com/manual/release-notes/3.6-compatibility/#remove-pushall-update-operator), so everyone downloading the latest and greatest mongo will run into this.

The solution here is to either

1) Use MongoDB V3.4.X (or 3.5 for that matter, but it's an unstable dev release)

OR

2) add `usePushEach: true` option to your models that will require this functionality.

I recommend #1 for production apps.

**Comment by shagunc96** [comment]
@knutole That solution worked fine for me. Using v3.6

```
var aSchema = new mongoose.Schema({
    ...
}, {
        usePushEach: true 
    }
);
```
Gives no errors.

**Comment by vkarpov15** [comment]
There's a fix for this in the upcoming 5.0.0-rc0 release, for now just use the usePushEach option.

**Comment by larsdecker** [comment]
Fix works in the 5.0.0-rc2 release.
